src/ui/i18n/i18n.js

```diff
@@ -0,0 +1,136 @@
+import path from 'path';
+import Promise from 'bluebird';
+import { readFile } from 'fs';
+import _ from 'lodash';
+
+const asyncReadFile = Promise.promisify(readFile);
+
+const TRANSLATION_FILE_EXTENSION = '.json';
+
+function getLocaleFromFileName(fullFileName) {
+  if (_.isEmpty(fullFileName)) throw new Error('Filename empty');
+
+  const fileExt = path.extname(fullFileName);
+  if (fileExt.length <= 0 || fileExt !== TRANSLATION_FILE_EXTENSION) {
+    throw new Error('Translations must be in a JSON file. File being registered is ' + fullFileName);
+  }
+
+  return path.basename(fullFileName, TRANSLATION_FILE_EXTENSION);
+}
+
+function getBestLocaleMatch(languageTag, registeredLocales) {
+  if (_.contains(registeredLocales, languageTag)) {
+    return languageTag;
+  }
+
+  // Find the first registered locale that begins with one of the language codes from the provided language tag.
+  // For example, if there is an 'en' language code, it would match an 'en-US' registered locale.
+  const languageCode = _.first(languageTag.split('-')) || [];
+  return _.find(registeredLocales, (locale) => _.startsWith(locale, languageCode));
+}
+
+export class I18n {
+
+  _registeredTranslations = {};
```

Comment:
I don't think we should rely on features that are only in stage 2 of the proposal process https://github.com/tc39/proposal-class-public-fields